File: ai-backend/app/notion_service.py
import logging
from functools import lru_cache

import httpx

from app.config import settings

logger = logging.getLogger(__name__)

NOTION_API_BASE = "https://api.notion.com/v1"
NOTION_VERSION = "2022-06-28"


def _headers() -> dict[str, str]:
    return {
        "Authorization": f"Bearer {settings.notion_api_key}",
        "Notion-Version": NOTION_VERSION,
        "Content-Type": "application/json",
    }


async def read_page(page_id: str) -> str:
    """Notion 페이지를 읽어 마크다운으로 반환."""
    async with httpx.AsyncClient(base_url=NOTION_API_BASE, headers=_headers(), timeout=30) as client:
        title = await _fetch_title(client, page_id)
        blocks = await _fetch_all_blocks(client, page_id)
        content = _blocks_to_markdown(blocks, depth=0)
        return f"# {title}\n\n{content}"


async def _fetch_title(client: httpx.AsyncClient, page_id: str) -> str:
    try:
        resp = await client.get(f"/pages/{page_id}")
        resp.raise_for_status()
        data = resp.json()
        properties = data.get("properties", {})
        for prop in properties.values():
            if prop.get("type") == "title":
                title_arr = prop.get("title", [])
                if title_arr:
                    return "".join(t.get("plain_text", "") for t in title_arr)
    except Exception as e:
        logger.warning("페이지 제목 추출 실패: %s", e)
    return "Untitled"


async def _fetch_all_blocks(client: httpx.AsyncClient, block_id: str) -> list[dict]:
    """커서 기반 페이지네이션으로 모든 블록 가져오기."""
    blocks = []
    cursor = None
    while True:
        params = {"page_size": 100}
        if cursor:
            params["start_cursor"] = cursor
        resp = await client.get(f"/blocks/{block_id}/children", params=params)
        resp.raise_for_status()
        data = resp.json()
        blocks.extend(data.get("results", []))

        if not data.get("has_more"):
            break
        cursor = data.get("next_cursor")

    # 자식 블록 재귀 fetch
    for block in blocks:
        if block.get("has_children"):
            block["_children"] = await _fetch_all_blocks(client, block["id"])

    return blocks


def _blocks_to_markdown(blocks: list[dict], depth: int) -> str:
    lines = []
    for block in blocks:
        line = _block_to_markdown(block, depth)
        if line:
            lines.append(line)
        children = block.get("_children", [])
        if children:
            lines.append(_blocks_to_markdown(children, depth + 1))
    return "\n".join(lines)


def _block_to_markdown(block: dict, depth: int) -> str:
    block_type = block.get("type", "")
    data = block.get(block_type, {})
    indent = "  " * depth

    match block_type:
        case "paragraph":
            return indent + _rich_text(data)
        case "heading_1":
            return f"# {_rich_text(data)}"
        case "heading_2":
            return f"## {_rich_text(data)}"
        case "heading_3":
            return f"### {_rich_text(data)}"
        case "bulleted_list_item":
            return f"{indent}- {_rich_text(data)}"
        case "numbered_list_item":
            return f"{indent}1. {_rich_text(data)}"
        case "to_do":
            checked = data.get("checked", False)
            marker = "- [x]" if checked else "- [ ]"
            return f"{indent}{marker} {_rich_text(data)}"
        case "toggle":
            return f"{indent}<details><summary>{_rich_text(data)}</summary></details>"
        case "quote":
            return f"{indent}> {_rich_text(data)}"
        case "code":
            lang = data.get("language", "")
            return f"{indent}```{lang}\n{indent}{_rich_text(data)}\n{indent}```"
        case "callout":
            icon = _callout_icon(data)
            return f"{indent}> {icon} {_rich_text(data)}"
        case "divider":
            return f"{indent}---"
        case "image":
            return f"{indent}{_media_url(data, 'image')}"
        case "bookmark":
            url = data.get("url", "")
            return f"{indent}[Bookmark]({url})"
        case _:
            return ""


def _rich_text(data: dict) -> str:
    texts = data.get("rich_text", [])
    parts = []
    for rt in texts:
        text = rt.get("plain_text", "")
        ann = rt.get("annotations", {})
        if ann.get("bold"):
            text = f"**{text}**"
        if ann.get("italic"):
            text = f"*{text}*"
        if ann.get("strikethrough"):
            text = f"~~{text}~~"
        if ann.get("code"):
            text = f"`{text}`"
        href = rt.get("href")
        if href:
            text = f"[{text}]({href})"
        parts.append(text)
    return "".join(parts)


def _callout_icon(data: dict) -> str:
    icon = data.get("icon", {})
    if icon.get("type") == "emoji":
        return icon.get("emoji", "")
    return ""


def _media_url(data: dict, media_type: str) -> str:
    url_type = data.get("type", "")
    url_data = data.get(url_type, {})
    url = url_data.get("url", "")
    return f"![{media_type}]({url})" if url else ""

File: ai-backend/src/main/java/com/portfolio/ai/service/NotionService.java
package com.portfolio.ai.service;

import com.github.benmanes.caffeine.cache.Cache;
import com.portfolio.ai.config.AiProperties;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.client.WebClientResponseException;
import org.springframework.web.server.ResponseStatusException;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

@Slf4j
@Service
public class NotionService {

    private static final String NOTION_API_BASE = "https://api.notion.com/v1";
    private static final String NOTION_VERSION = "2022-06-28";
    private static final int MAX_CONCURRENCY = 5;

    private final WebClient webClient;
    private final Cache<String, String> notionPageCache;

    public NotionService(WebClient.Builder webClientBuilder,
                         AiProperties aiProperties,
                         Cache<String, String> notionPageCache) {
        this.webClient = webClientBuilder
                .baseUrl(NOTION_API_BASE)
                .defaultHeader(HttpHeaders.AUTHORIZATION, "Bearer " + aiProperties.getNotionApiKey())
                .defaultHeader("Notion-Version", NOTION_VERSION)
                .defaultHeader(HttpHeaders.CONTENT_TYPE, "application/json")
                .build();
        this.notionPageCache = notionPageCache;
    }

    public Mono<String> readPage(String pageId) {
        String cached = notionPageCache.getIfPresent(pageId);
        if (cached != null) {
            log.debug("Notion 페이지 캐시 히트: {}", pageId);
            return Mono.just(cached);
        }

        log.info("Notion 페이지 조회 시작: {}", pageId);

        Mono<String> titleMono = fetchPageTitle(pageId);
        Mono<String> contentMono = fetchAllBlocks(pageId);

        return Mono.zip(titleMono, contentMono)
                .map(tuple -> {
                    String title = tuple.getT1();
                    String content = tuple.getT2();
                    String markdown = "# " + title + "\n\n" + content;
                    notionPageCache.put(pageId, markdown);
                    log.info("Notion 페이지 조회 완료: {} ({}자)", pageId, markdown.length());
                    return markdown;
                })
                .onErrorMap(WebClientResponseException.NotFound.class,
                        e -> new ResponseStatusException(HttpStatus.NOT_FOUND,
                                "Notion 페이지를 찾을 수 없습니다: " + pageId))
                .onErrorMap(WebClientResponseException.Unauthorized.class,
                        e -> new ResponseStatusException(HttpStatus.UNAUTHORIZED,
                                "Notion API 인증에 실패했습니다. API 키를 확인해주세요."));
    }

    private Mono<String> fetchPageTitle(String pageId) {
        return webClient.get()
                .uri("/pages/{pageId}", pageId)
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<Map<String, Object>>() {})
                .map(this::extractTitle)
                .onErrorReturn("Untitled");
    }

    @SuppressWarnings("unchecked")
    private String extractTitle(Map<String, Object> pageData) {
        try {
            Map<String, Object> properties = (Map<String, Object>) pageData.get("properties");
            if (properties == null) return "Untitled";

            for (Map.Entry<String, Object> entry : properties.entrySet()) {
                Map<String, Object> prop = (Map<String, Object>) entry.getValue();
                if ("title".equals(prop.get("type"))) {
                    List<Map<String, Object>> titleArray = (List<Map<String, Object>>) prop.get("title");
                    if (titleArray != null && !titleArray.isEmpty()) {
                        return titleArray.stream()
                                .map(t -> (String) t.get("plain_text"))
                                .collect(Collectors.joining());
                    }
                }
            }
        } catch (Exception e) {
            log.warn("페이지 제목 추출 실패: {}", e.getMessage());
        }
        return "Untitled";
    }

    private Mono<String> fetchAllBlocks(String pageId) {
        return fetchBlockChildren(pageId)
                .flatMap(blocks -> processBlocksRecursively(blocks, 0))
                .map(indexedContents -> indexedContents.stream()
                        .sorted(Comparator.comparingInt(IndexedContent::index))
                        .map(IndexedContent::content)
                        .collect(Collectors.joining("\n")));
    }

    @SuppressWarnings("unchecked")
    private Mono<List<Map<String, Object>>> fetchBlockChildren(String blockId) {
        return webClient.get()
                .uri("/blocks/{blockId}/children?page_size=100", blockId)
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<Map<String, Object>>() {})
                .expand(response -> {
                    Boolean hasMore = (Boolean) response.get("has_more");
                    if (Boolean.TRUE.equals(hasMore)) {
                        String nextCursor = (String) response.get("next_cursor");
                        return webClient.get()
                                .uri("/blocks/{blockId}/children?page_size=100&start_cursor={cursor}",
                                        blockId, nextCursor)
                                .retrieve()
                                .bodyToMono(new ParameterizedTypeReference<Map<String, Object>>() {});
                    }
                    return Mono.empty();
                })
                .flatMapIterable(response -> {
                    List<Map<String, Object>> results = (List<Map<String, Object>>) response.get("results");
                    return results != null ? results : Collections.emptyList();
                })
                .collectList();
    }

    private Mono<List<IndexedContent>> processBlocksRecursively(List<Map<String, Object>> blocks, int depth) {
        AtomicInteger index = new AtomicInteger(0);

        return Flux.fromIterable(blocks)
                .map(block -> new BlockWithIndex(block, index.getAndIncrement()))
                .flatMap(bwi -> {
                    String content = convertBlockToMarkdown(bwi.block(), depth);
                    Mono<List<IndexedContent>> childrenMono;

                    Boolean hasChildren = (Boolean) bwi.block().get("has_children");
                    if (Boolean.TRUE.equals(hasChildren)) {
                        String blockId = (String) bwi.block().get("id");
                        childrenMono = fetchBlockChildren(blockId)
                                .flatMap(children -> processBlocksRecursively(children, depth + 1));
                    } else {
                        childrenMono = Mono.just(Collections.emptyList());
                    }

                    return childrenMono.map(children -> {
                        List<IndexedContent> result = new ArrayList<>();
                        result.add(new IndexedContent(bwi.index(), content));
                        for (int i = 0; i < children.size(); i++) {
                            IndexedContent child = children.get(i);
                            result.add(new IndexedContent(
                                    bwi.index() * 1000 + child.index() + 1,
                                    child.content()));
                        }
                        return result;
                    });
                }, MAX_CONCURRENCY)
                .flatMapIterable(list -> list)
                .collectList();
    }

    @SuppressWarnings("unchecked")
    private String convertBlockToMarkdown(Map<String, Object> block, int depth) {
        String type = (String) block.get("type");
        if (type == null) return "";

        String indent = "  ".repeat(depth);
        Map<String, Object> typeData = (Map<String, Object>) block.get(type);

        return switch (type) {
            case "paragraph" -> indent + extractRichText(typeData);
            case "heading_1" -> "# " + extractRichText(typeData);
            case "heading_2" -> "## " + extractRichText(typeData);
            case "heading_3" -> "### " + extractRichText(typeData);
            case "bulleted_list_item" -> indent + "- " + extractRichText(typeData);
            case "numbered_list_item" -> indent + "1. " + extractRichText(typeData);
            case "toggle" -> indent + "<details><summary>" + extractRichText(typeData) + "</summary></details>";
            case "quote" -> indent + "> " + extractRichText(typeData);
            case "code" -> {
                String language = typeData != null ? (String) typeData.get("language") : "";
                yield indent + "```" + (language != null ? language : "") + "\n"
                        + indent + extractRichText(typeData) + "\n"
                        + indent + "```";
            }
            case "callout" -> {
                String icon = extractCalloutIcon(typeData);
                yield indent + "> " + icon + " " + extractRichText(typeData);
            }
            case "divider" -> indent + "---";
            case "to_do" -> {
                boolean checked = typeData != null && Boolean.TRUE.equals(typeData.get("checked"));
                yield indent + (checked ? "- [x] " : "- [ ] ") + extractRichText(typeData);
            }
            case "child_page" -> {
                String title = typeData != null ? (String) typeData.get("title") : "Untitled";
                yield indent + "**" + title + "**";
            }
            case "child_database" -> {
                String title = typeData != null ? (String) typeData.get("title") : "Untitled";
                yield indent + "**" + title + "**";
            }
            case "image" -> indent + extractMediaUrl(typeData, "image");
            case "bookmark" -> {
                String url = typeData != null ? (String) typeData.get("url") : "";
                yield indent + "[Bookmark](" + url + ")";
            }
            case "link_preview" -> {
                String url = typeData != null ? (String) typeData.get("url") : "";
                yield indent + "[Link](" + url + ")";
            }
            case "table_of_contents" -> indent + "[TOC]";
            default -> {
                log.debug("지원하지 않는 블록 타입: {}", type);
                yield "";
            }
        };
    }

    @SuppressWarnings("unchecked")
    private String extractRichText(Map<String, Object> typeData) {
        if (typeData == null) return "";
        List<Map<String, Object>> richTexts = (List<Map<String, Object>>) typeData.get("rich_text");
        if (richTexts == null || richTexts.isEmpty()) return "";

        return richTexts.stream()
                .map(rt -> {
                    String text = (String) rt.get("plain_text");
                    if (text == null) return "";
                    Map<String, Object> annotations = (Map<String, Object>) rt.get("annotations");
                    if (annotations != null) {
                        if (Boolean.TRUE.equals(annotations.get("bold"))) text = "**" + text + "**";
                        if (Boolean.TRUE.equals(annotations.get("italic"))) text = "*" + text + "*";
                        if (Boolean.TRUE.equals(annotations.get("strikethrough"))) text = "~~" + text + "~~";
                        if (Boolean.TRUE.equals(annotations.get("code"))) text = "`" + text + "`";
                    }
                    String href = (String) rt.get("href");
                    if (href != null) text = "[" + text + "](" + href + ")";
                    return text;
                })
                .collect(Collectors.joining());
    }

    @SuppressWarnings("unchecked")
    private String extractCalloutIcon(Map<String, Object> typeData) {
        if (typeData == null) return "";
        Map<String, Object> icon = (Map<String, Object>) typeData.get("icon");
        if (icon == null) return "";
        String iconType = (String) icon.get("type");
        if ("emoji".equals(iconType)) return (String) icon.get("emoji");
        return "";
    }

    @SuppressWarnings("unchecked")
    private String extractMediaUrl(Map<String, Object> typeData, String mediaType) {
        if (typeData == null) return "";
        String urlType = (String) typeData.get("type");
        if (urlType == null) return "";
        Map<String, Object> urlData = (Map<String, Object>) typeData.get(urlType);
        if (urlData == null) return "";
        String url = (String) urlData.get("url");
        if (url == null) return "";
        return "![" + mediaType + "](" + url + ")";
    }

    record BlockWithIndex(Map<String, Object> block, int index) {}
    record IndexedContent(int index, String content) {}
}
